posix: add networking apis

cfriedt opened the pull request:
Add implementations (some fully supported, some stubs) for the remaining POSIX networking API calls
```
endhostent
endnetent
endprotoent
endservent
gethostent
getnetbyaddr
getnetbyname
getnetent
getprotobyname
getprotobynumber
getprotoent
getservbyname
getservbyport
getservent
if_freenameindex
if_indextoname
if_nameindex
if_nametoindex
inet_addr
inet_ntoa
sethostent
setnetent
setprotoent
setservent
sockatmark
```

With this, and with existing network subsystem API, we have mostly complete support for the `POSIX_NETWORKING` Option Group, as described in Section E.1 of IEEE 1003.1-2017.

The `POSIX_NETWORKING` Option Group is required by PSE53 and PSE54 Application Environment Profiles.

cfriedt committed:
posix: net: add kconfig option and inet_addr()

Add a Kconfig option for POSIX_NETWORKING and implement
inet_addr().

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
posix: net: implement inet_ntoa()

Add an implementation of inet_ntoa().

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: posix: net: tests for inet_ntoa()

Add tests for inet_ntoa()

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: posix: net: add tests for if_nameindex et al

Add tests for if_indextoname(), if_freenameindex(),
if_nameindex(), and if_nametoindex().

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
posix: net: implement if_nameindex() et al

Implement if_indextoname(), if_freenameindex(), if_nameindex(),
and if_nametoindex() by wrapping around networking subystem
calls.

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: posix: net: add a posix network api testsuite

Add a testsuite for posix networking apis with an initial
test for inet_addr().

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
posix: net: add trivial netdb function implementations

Add trivial function implementations for those declared in
netdb.h .

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: posix: headers: uncomment netdb.h tests

Uncomment netdb.h tests now that we have an implementation and
can check for the existence of e.g. hostent, netent, protoent,
and servent structure fields, as well as associated accessor
functions.

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: posix: headers: uncomment net/if.h tests

Uncomment net/if.h tests in tests/posix/headers since we now
have an implementation and can check for the existence of
structures, functions, etc.

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
posix: net: add a trivial implementation of sockatmark()

Add a trivial implementation of sockatmark().

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: posix: headers: uncomment check for sockatmark()

Uncomment the check for sockatmark() to ensure that the function
is declared in sys/socket.h, as a stub implementation was added
in the previous commit.

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

cfriedt committed:
doc: posix: add POSIX_NETWORKING Option Group and mark supported

Mark the POSIX_NETWORKING Option Group as supported. The
POSIX_NETWORKING Option Group is required for PSE53 Conformance.

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

jukkar reviewed (dismissed):
Thanks Chris, looks very good :bowing_man:

cfriedt commented on tests/posix/net/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+    - net
+    - posix
+  depends_on: netif
+  arch_exclude:
```
I found this was necessary because `native_sim` requires super user privileges (at least on my desk).

Can be fixed later, most likely

cfriedt commented:
Had to add a cast to `int` in calls to `isdigit()` and `isspace()` for newlib to avoid these warnings being promoted to errors.
```
/__w/zephyr/zephyr/lib/posix/options/net.c: In function 'inet_addr':
/__w/zephyr/zephyr/lib/posix/options/net.c:41:38: error: array subscript has type 'char' [-Werror=char-subscripts]
   41 |                 } else if (isdigit(cp[i])) {
      |                                    ~~^~~
/__w/zephyr/zephyr/lib/posix/options/net.c:45:38: error: array subscript has type 'char' [-Werror=char-subscripts]
   45 |                 } else if (isspace(cp[i])) {
      |                                    ~~^~~
cc1: all warnings being treated as errors
```

cfriedt committed:
tests: posix: common: reorder headers to avoid LOG_ERR conflict

The log subsystem defines LOG_ERR as a macro, which conflicts
with the standard POSIX LOG_ERR log level (3).

Reorder includes to avoid errors.

Signed-off-by: Chris Friedt <chrisfriedt@gmail.com>

jukkar commented on tests/posix/headers/src/net_if_h.c:
```diff
@@ -19,15 +19,15 @@
  */
```
About the commit `tests: posix: headers: uncomment net/if.h tests`,
could you add a slightly more informative commit text, the current one states the obvious, but what should be mentioned here why we need/want to uncomment the test.
Same comment to previous commit about the netdb.h

cfriedt replied:
Fixed!

cfriedt commented:
This is running fine locally with `net-setup.sh`  from `net-tools` first and then `west build -p auto -b qemu_x86 -t run tests/posix/net`. Twister runs fine locally too with `twister -i -p qemu_x86 -T tests/posix/net/`

I'm not entirely sure what is up with CI.

```shell
*** Booting Zephyr OS build v3.6.0-1560-g064b91cf8792 ***
Running TESTSUITE net
===================================================================
START - test_if_freenameindex
 PASS - test_if_freenameindex in 0.001 seconds
===================================================================
START - test_if_indextoname
1 interfaces
interface 1: eth0
 PASS - test_if_indextoname in 0.001 seconds
===================================================================
START - test_if_nameindex
1 interfaces
interface 1: eth0
 PASS - test_if_nameindex in 0.001 seconds
===================================================================
START - test_if_nametoindex
1 interfaces
interface 1: eth0
 PASS - test_if_nametoindex in 0.001 seconds
===================================================================
START - test_inet_addr
 PASS - test_inet_addr in 0.001 seconds
===================================================================
START - test_inet_ntoa
 PASS - test_inet_ntoa in 0.001 seconds
===================================================================
TESTSUITE net succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [net]: pass = 6, fail = 0, skip = 0, total = 6 duration = 0.006 seconds
 - PASS - [net.test_if_freenameindex] duration = 0.001 seconds
 - PASS - [net.test_if_indextoname] duration = 0.001 seconds
 - PASS - [net.test_if_nameindex] duration = 0.001 seconds
 - PASS - [net.test_if_nametoindex] duration = 0.001 seconds
 - PASS - [net.test_inet_addr] duration = 0.001 seconds
 - PASS - [net.test_inet_ntoa] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

pdgendt commented on tests/posix/net/src/inet_addr.c:
```diff
@@ -0,0 +1,50 @@
+
+	ARRAY_FOR_EACH_PTR(parms, p) {
+		ret = inet_addr(p->in);
+		zassert_equal(ret, p->out, "inet_addr(%s) failed. expect: %d actual: %d", p->in,
+			      p->out, ret);
```
A general remark for loops in tests, I think it's more helpful to use the `zexpect_` variants, which also fails the test, but can show all the failing cases, instead of stopping on the first one.

This can already reveal a pattern to be fixed, which might not be obvious from a single failed case.

cfriedt replied:
Fixed!

rlubos commented on lib/posix/options/net.c:
```diff
@@ -275,3 +275,9 @@ void setservent(int stayopen)
+
+int sockatmark(int s)
+{
+	errno = ENOSYS;
```
nit: `ARG_UNUSED(s);`
